modules/_labs/select/react/lib/Select.tsx

```diff
@@ -0,0 +1,342 @@
+    });
+
+    // offer visual feedback (flashing animation) briefly
+    // before dismissing menu; set new label when menu has been
+    // dismissed
+    setTimeout(() => {
+      this.setState({
+        justSelectedItemIndex: null,
+        label: optionProps.label,
+      });
+      this.toggleMenu(false);
+    }, dismissMenuTimeout);
+
+    if (this.inputRef && this.inputRef.current) {
+      const nativeInputValue = Object.getOwnPropertyDescriptor(
+        Object.getPrototypeOf(this.inputRef.current),
+        'value'
+      );
+      if (nativeInputValue && nativeInputValue.set) {
+        nativeInputValue.set.call(this.inputRef.current, optionProps.value);
+      }
+
+      let event: Event;
+      if (typeof Event === 'function') {
+        // modern browsers
+        event = new Event('input', {bubbles: true});
+      } else {
+        // IE 11
+        event = document.createEvent('Event');
+        event.initEvent('input', true);
+      }
+
+      this.inputRef.current.dispatchEvent(event);
+    }
+  };
+
+  handleKeyboardShortcuts = (event: React.KeyboardEvent): void => {
+    // console.log(event.key);
+
+    const children = React.Children.toArray(this.props.children);
+    const itemCount = children.length;
+    let isShortcut = false;
+    let nextFocusedIndex = 0;
+
+    switch (event.key) {
+      case 'ArrowUp':
+      case 'ArrowDown':
+        if (!this.state.showingMenu) {
+          this.toggleMenu(true);
+        } else if (this.state.focusedItemIndex !== null) {
+          const direction = event.key === 'ArrowUp' ? -1 : 1;
+          isShortcut = true;
+          let nextIndex = this.state.focusedItemIndex + direction;
+          while (nextIndex < itemCount && nextIndex >= 0 && children[nextIndex].props.disabled) {
+            nextIndex += direction;
+          }
+          nextFocusedIndex = nextIndex < 0 ? 0 : nextIndex >= itemCount ? itemCount - 1 : nextIndex;
+          this.setState({focusedItemIndex: nextFocusedIndex});
+        }
+        break;
+
+      case 'Spacebar':
+      case ' ':
+      case 'Enter':
+        if (this.state.focusedItemIndex !== null) {
+          const child = children[this.state.focusedItemIndex] as React.ReactElement<
+            SelectOptionProps
+          >;
+          this.handleOptionClick(child.props);
+          isShortcut = true;
+        }
+        break;
+
+      default:
+    }
+
+    if (isShortcut) {
+      event.stopPropagation();
+      event.preventDefault();
+    }
+  };
+
+  renderChildren = (
+    child: React.ReactElement<SelectOptionProps>,
+    index: number
+  ): React.ReactNode => {
+    return React.cloneElement(child, {
+      focused: this.state.focusedItemIndex === index,
+      justSelected: this.state.justSelectedItemIndex === index,
+      onMouseDown: (event: React.MouseEvent) => {
+        event.preventDefault();
+        this.handleOptionClick(child.props);
+      },
+      selected: this.state.selectedItemIndex === index,
+    });
```

Comment:
I remember @irobato using Context instead of `React.cloneElement` for various reasons. One of them is cloning elements limits how options can be passed to the Select.

The API as it is now makes the assumption that there will not be many options and that they will always be known. What about larger selects that might need dynamic options? Maybe the `Select` component isn't a good component for those use-cases, so maybe this is fine?

Reply:
Resolved now that we're using a render prop instead of cloneElement